src/main/java/roomescape/application/config/PaymentClientConfig.java

```diff
@@ -1,35 +1,52 @@
 package roomescape.application.config;
 
 import java.time.Duration;
-import org.springframework.beans.factory.annotation.Value;
-import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.web.client.ClientHttpRequestFactories;
+import org.springframework.boot.web.client.ClientHttpRequestFactorySettings;
+import org.springframework.boot.web.client.RestClientCustomizer;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
-import org.springframework.web.client.ResponseErrorHandler;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.http.client.JdkClientHttpRequestFactory;
 import org.springframework.web.client.RestClient;
-import org.springframework.web.client.RestTemplate;
-import roomescape.application.payment.PaymentErrorHandler;
+import roomescape.application.payment.PaymentClient;
+import roomescape.infrastructure.payment.TossPaymentClient;
 
 @Configuration
+@EnableConfigurationProperties(PaymentClientProperties.class)
 public class PaymentClientConfig {
-    private final String paymentUrl;
+    private final PaymentClientProperties properties;
 
-    public PaymentClientConfig(@Value("${payment.url}") String paymentUrl) {
-        this.paymentUrl = paymentUrl;
+    public PaymentClientConfig(PaymentClientProperties properties) {
+        this.properties = properties;
     }
 
     @Bean
-    public ResponseErrorHandler errorHandler() {
-        return new PaymentErrorHandler();
+    public RestClientCustomizer paymentRestClientCustomizer() {
```

Comment:
이 부분은 테스트를 용이하게 하기 위해서 등장했습니다.
`RestClientTest`에서는 `MockRestServiceServer`를 `Builder`를 통해 추적합니다. 이 과정에서 주입받은 `builder`에 `.requestFactory()`를 사용하면 추적 과정이 오버라이딩돼 더이상 Mock이 되지 않습니다 ([Reference](https://github.com/spring-projects/spring-boot/issues/38832))

> Calling RestClient.Builder.requestFactory() directly is overriding this and breaking the mocking. You can work around this for now by using a RestClientCustomizer instead of directly modifying the builder.

다만 이 경우에는 한 번에 여러 base-url에 대한 대처가 어려워질 듯합니다. 전역으로 설정되어버리는 느낌이 있어서 그런데요, 조금 더 유연하게 관리하는 방법이 있을까요?

Reply:
제가 질문을 제대로 이해한 것인지 모르겠는데요.

base url이 여러개가 될때 이 customizer를 사용하기 힘들어져서 그런것일까요?

저는 보통 customizer는 사용하지않고 RestClient.Builder를 여러개 생성하고, RestClient.Builder의 컬렉션을 가진 객체를 Bean으로 생성하도록 합니다.

이렇게 customizer를 사용하면 전역 customizer로 설정이되어서 세세한 설정을 각 결제사 별로 적용하기가 힘들어서요.

조금 더 자세히 설명해보면

보통 application.yml에 base-url과 각 결제사/환경 별 conn/read timeout을 기재해두는데요.

이를 리스트 형태로 불러와서 빌더를 여러개 생성합니다.

테스트를 할때 mockRestServiceServer를 직접 생성하는 편이기도 합니다. 이때 인자로 Builder가 필요하기도 해서 위와 같이 사용하는 편입니다.

Reply:
> 이를 리스트 형태로 불러와서 빌더를 여러개 생성합니다.

와 빌더 일급 컬렉션으로 빈 등록할 생각을 못 했네요. Bean과 Component에 대한 시각이 넓어졌습니다. 단순한 객체에 스프링 제어가 추가되니 마법이 시작되네요 👀

추가해 보았는데요, 구조가 괜찮을지는 모르겠습니다. 피드백 받아가면서 발전해나가면 좋겠네요 🙂